keps/sig-node/kubelet-node-endpoint.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Kubelet Node Endpoint
+authors:
+  - "clamoriniere"
+owning-sig: sig-node
+participating-sigs: []
+reviewers: []
+approvers: []
+creation-date: 2019-05-02
+status: implementable
+---
+
+# Kubelet Node Endpoint
+
+## Motivation
+
+The Kubelet already proposes a set of endpoints (`/pods`, and soon `/metrics/resource/v1*`) that offer useful information for "local" node monitoring. In addition to those endpoints, we would like to introduce a new `/node` endpoint for displaying the node resource information. It will allow daemonset-based monitoring solutions to retrieve node information locally (on the same host) like "labels, annotations" instead of using a call to the API server, which will solve some scalability issues in big clusters.
+
+![kubelet-node-endpoint](https://i.ibb.co/khVg8rd/Untitled-Diagram.png)
+
+## Proposal
+
+Kubelet already caches and sync its corresponding `Node` information from the API-server. The proposed solution opens a new endpoint "/node" that returns the `k8s.io/api/pkg/core/v1.Node` JSON representation of the Kubelet's Node. Only the Kubelet's Node resource is return.
+
+For consistency the endpoint path "/node" looks similar to the "/pods" endpoint (no API versioning).
+
+## Use Cases
+
+* **Pod metrics and logs tagging with Node labels or annotation**: Monitoring workload solutions based on the kubelet API ("/pods" and "/metrics") used to decorate Pod metrics with additional context information; Information like the "Availability zone", "Provider", "Server type" etc are very useful for issue investigation.
+The current workaround is to retrieve that information from the API-Server, that generates a consequent number of requests to the API-Server on a large kubernetes cluster.
+* **Container images monitoring on Node**: Knowing which Images and Disk usage that it represents on a specific node can be useful for Node monitoring.
```

Comment:
+1
I would add that we need per-layer image accounting, rather than per-image.  Simply summing the total size of all images, as we do today in the node status gives a result that is larger than the actual amount of disk space used by images.

Regardless of whether we end up exposing the node object locally, this is a monitoring gap we need to fill.

Reply:
We also cap the number of images in the node status, which makes the proposed solution less appealing to solve this particular problem.

Reply:
I didn't know about the images in the node status may not be accurate.
And agree with you, having the information with an image layers granularity will be more useful.

Reply:
@dashpole do you want that I create another KEP for the container images disk usage?

Reply:
If you have good ideas for solving the issue, then sure.  It isn't immediately obvious to me what the best way to represent per-image-layer metrics is in a way that relates back to containers and pods.  What metric would an application developer actually want to see WRT their image?